File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        
        def find(node, left, right):
            if node is None:
                return None
            
            if left<=node.val<=right:
                return node
            
            if right<node.val:
                return find(node.left,left,right)
            return find(node.right,left,right)
            
        return find(root,min(p.val,q.val),max(p.val,q.val))

File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lca(TreeNode node, int min, int max){
        if(node==null)return node;
        if(min<=node.val && node.val<=max)return node;
        if(max<node.val)return lca(node.left,min,max);
        return lca(node.right,min,max);
    }
    
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        return lca(root, Math.min(p.val,q.val), Math.max(p.val,q.val));
    }
}